src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml

```diff
@@ -107,11 +107,17 @@ services:
     calls:
       - [ "setEmployeeId", [ "@=service('prestashop.adapter.legacy.context').getContext().employee?service('prestashop.adapter.legacy.context').getContext().employee.id:0" ] ]
 
-  prestashop.categories_provider:
-    class: PrestaShopBundle\Service\DataProvider\Admin\CategoriesProvider
+  prestashop.categories_provider_builder:
+    class: PrestaShopBundle\Service\DataProvider\Admin\CategoriesProviderBuilder
     arguments:
+      - '@prestashop.adapter.legacy.context'
+      - '@prestashop.core.addon.theme.repository'
       - "%prestashop.addons.categories%"
-      - "@=service('prestashop.adapter.legacy.context').getContext().shop.theme_name?service('prestashop.core.addon.theme.repository').getInstanceByName(service('prestashop.adapter.legacy.context').getContext().shop.theme_name).getModulesToEnable():[]"
+
+  prestashop.categories_provider:
+    lazy: true
```

Comment:
We don't want to instantiate it directly when the application starts, the context is not initialized when executing a command during the installation phase. https://github.com/PrestaShop/PrestaShop/issues/29557#issuecomment-1539676702 for more info

Reply:
Yes, and too many things were executed in the service definition. We wanted to refactor the `CategoriesProvider` class so that it depends on the context and repo in order to execute the fetching part later only when needed

The problem is that this class is used in other use cases and needs to be constructed with an array and not a service like this case:

https://github.com/PrestaShop/PrestaShop/blob/455a2cbf40ef887a81367a32520955e94f7ad330/install-dev/controllers/http/content.php#L139

So to keep the compatibility with other usage we thought about introducing a factory that replaces the code executed via yaml definition. And to prevent the factory from being called too soon a lazy parameter was also needed It allows to delay the creation of the service later when the context is initialized

I admit it's not the best/ideal solution but it's acceptable for a fix IMHO